File: product-processor-service/src/main/java/com/example/product_processor_service/service/marketplace/MarketplaceServiceManager.java
package com.example.product_processor_service.service.marketplace;

import com.example.product_processor_service.repository.marketplace.definition.path.MarketplacePathDefinitionRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class MarketplaceServiceManager implements MarketplaceService {
    private final MarketplacePathDefinitionRepository marketplacePathDefinitionRepository;

    public MarketplaceServiceManager(MarketplacePathDefinitionRepository marketplacePathDefinitionRepository) {
        this.marketplacePathDefinitionRepository = marketplacePathDefinitionRepository;
    }

    @Override
    public boolean isMarketplaceSupported(String marketplaceDomain) {
        if(marketplaceDomain.isBlank()) {
            throw new IllegalArgumentException("marketplaceDomain cannot be blank");
        }
        return marketplacePathDefinitionRepository.existsByPath(marketplaceDomain);
    }

    @Override
    public List<String> getSupportedMarketplaces() {
        return marketplacePathDefinitionRepository.getAllPaths();
    }
}

File: product-manager-service/src/main/kotlin/com/example/product_manager_service/service/marketplace/MarketplaceServiceManager.kt
package com.example.product_manager_service.service.marketplace

import com.example.product_manager_service.service.common.grpc.mapper.GrpcMapperService
import com.example.product_manager_service.service.marketplace.grpc.MarketplaceGrpcClientService
import org.springframework.stereotype.Service

@Service
class MarketplaceServiceManager(private val marketplaceGrpcClientService: MarketplaceGrpcClientService): MarketplaceService {
    private lateinit var supportedMarketplaces: Set<String>

    override fun getSupported(): Set<String> {
        if(!this::supportedMarketplaces.isInitialized) {
            supportedMarketplaces = GrpcMapperService.mapTo(marketplaceGrpcClientService.getSupportedMarketplaces())
        }
        return supportedMarketplaces
    }
}